src/ledger/database/lmdb.zig

```diff
@@ -0,0 +1,716 @@
+            return struct {
+                allocator: Allocator,
+                txn: *c.MDB_txn,
+                cursor: *c.MDB_cursor,
+                direction: CursorRelativeOperation,
+                next_operation: CursorRelativeOperation,
+
+                /// Calling this will free all slices returned by the iterator
+                pub fn deinit(self: *@This()) void {
+                    c.mdb_cursor_close(self.cursor);
+                    c.mdb_txn_abort(self.txn);
+                }
+
+                pub fn next(self: *@This()) anyerror!?cf.Entry() {
+                    const key, const val = try self.nextImpl() orelse return null;
+                    return .{
+                        try key_serializer.deserialize(cf.Key, self.allocator, key),
+                        try value_serializer.deserialize(cf.Value, self.allocator, val),
+                    };
+                }
+
+                pub fn nextKey(self: *@This()) anyerror!?cf.Key {
+                    const key, _ = try self.nextImpl() orelse return null;
+                    return try key_serializer.deserialize(cf.Key, self.allocator, key);
+                }
+
+                pub fn nextValue(self: *@This()) anyerror!?cf.Value {
+                    _, const val = try self.nextImpl() orelse return null;
+                    return try key_serializer.deserialize(cf.Value, self.allocator, val);
+                }
+
+                /// Returned data does not outlive the iterator.
+                pub fn nextBytes(self: *@This()) LmdbError!?[2]BytesRef {
+                    const key, const val = try self.nextImpl() orelse return null;
+                    return .{
+                        .{ .allocator = null, .data = key },
+                        .{ .allocator = null, .data = val },
+                    };
+                }
+
+                fn nextImpl(self: *@This()) LmdbError!?struct { []const u8, []const u8 } {
+                    defer self.next_operation = self.direction;
+                    return try cursorGetRelative(self.cursor, self.next_operation);
+                }
+            };
+        }
+    };
+}
+
+fn toVal(bytes: []const u8) c.MDB_val {
+    return .{
+        .mv_size = bytes.len,
+        .mv_data = @constCast(@ptrCast(bytes.ptr)),
+    };
+}
+
+fn fromVal(value: c.MDB_val) []const u8 {
+    const ptr: [*c]u8 = @ptrCast(value.mv_data);
+    return ptr[0..value.mv_size];
+}
+
+fn txnResetter(txn: *c.MDB_txn) Allocator {
+    const vtable = .{
+        .alloc = &sig.utils.allocators.noAlloc,
+        .resize = &Allocator.noResize,
+        .free = &resetTxnFree,
+    };
+    return .{
+        .ptr = @ptrCast(@alignCast(txn)),
+        .vtable = &vtable,
+    };
+}
+
+fn resetTxnFree(ctx: *anyopaque, _: []u8, _: u8, _: usize) void {
+    const txn: *c.MDB_txn = @ptrCast(@alignCast(ctx));
+    c.mdb_txn_abort(txn);
+}
+
+fn ret(constructor: anytype, args: anytype) LmdbError!TypeToCreate(constructor) {
```

Comment:
This, and the functions under it are clearly relying on some behavior that makes no sense to me, a casual reader. 
Could you document them better?

Reply:
f086a562142cb7b76e1bdfcf93ebed7f3e742c17